util/gctuner/tuner.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gctuner
+
+import (
+	"math"
+	"os"
+	"strconv"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/util"
+)
+
+const (
+	// MaxGCPercent is the default max cost of memory.
+	MaxGCPercent uint32 = 500
+	// MinGCPercent is the default min cost of memory.
+	MinGCPercent uint32 = 20
+)
+
+var defaultGCPercent uint32 = 100
+
+// EnableGOGCTuner is to control whether enable the GOGC tuner.
+var EnableGOGCTuner atomic.Bool
+
+func init() {
+	if val, err := strconv.Atoi(os.Getenv("GOGC")); err == nil {
+		defaultGCPercent = uint32(val)
+	}
+}
+
+// Tuning sets the threshold of heap which will be respect by gc tuner.
+// When Tuning, the env GOGC will not be take effect.
+// threshold: disable tuning if threshold == 0
+func Tuning(threshold uint64) {
+	// disable gc tuner if percent is zero
+	if threshold <= 0 && globalTuner != nil {
+		globalTuner.stop()
+		globalTuner = nil
+		return
+	}
+
+	if globalTuner == nil {
+		globalTuner = newTuner(threshold)
+		return
+	}
+	globalTuner.setThreshold(threshold)
+}
+
+// GetGOGC returns the current GCPercent.
+func GetGOGC() uint32 {
+	if globalTuner == nil {
+		return defaultGCPercent
+	}
+	return globalTuner.getGCPercent()
+}
+
+// only allow one gc tuner in one process
+var globalTuner *tuner
```

Comment:
The variable is not concurrent-safe, right? Maybe need some comments.

Reply:
add some comments.

```
// It is not thread-safe. so it is private to avoid misuse.
```